arches_references/datatypes/datatypes.py

```diff
@@ -22,66 +67,60 @@ def validate(
         strict=False,
         **kwargs,
     ):
-        errors = []
-        title = _("Invalid Reference Datatype Value")
-        if value is None:
-            return errors
-
-        if type(value) == list and len(value):
-            for reference in value:
-                if "uri" in reference and len(reference["uri"]):
-                    pass
-                else:
-                    errors.append(
-                        {
-                            "type": "ERROR",
-                            "message": _(
-                                "Reference objects require a 'uri' property and corresponding value"
-                            ),
-                            "title": title,
-                        }
-                    )
-                if "labels" in reference:
-                    pref_label_languages = []
-                    for label in reference["labels"]:
-                        if not all(
-                            key in label
-                            for key in ("id", "value", "language_id", "valuetype_id")
-                        ):
-                            errors.append(
-                                {
-                                    "type": "ERROR",
-                                    "message": _(
-                                        "Reference labels require properties: id(uuid), value(string), language_id(e.g. 'en'), and valuetype_id(e.g. 'prefLabel')"
-                                    ),
-                                    "title": title,
-                                }
-                            )
-                        if label["valuetype_id"] == "prefLabel":
-                            pref_label_languages.append(label["language_id"])
-
-                    if len(set(pref_label_languages)) < len(pref_label_languages):
-                        errors.append(
-                            {
-                                "type": "ERROR",
-                                "message": _(
-                                    "A reference can have only one prefLabel per language"
-                                ),
-                                "title": title,
-                            }
-                        )
-        else:
-            errors.append(
-                {
-                    "type": "ERROR",
-                    "message": _("Reference value must be a list of reference objects"),
-                    "title": title,
-                }
-            )
-        return errors
+        try:
+            parsed = self.to_python(value)
+            self.validate_pref_labels(parsed)
+            self.validate_multivalue(parsed, node, nodeid)
+        except Exception as e:
```

Comment:
should exceptions be caught/transformed at this level? Would it be a better pattern to do this in at the callsite?

Reply:
For better or worse I think this is the pattern used across datatypes in Arches. If something goes wrong, validate returns a list of errors. That list gets passed on to importers so that they can report everything that's wrong with whatever data a user is trying to load.